source/generators/MuonAngleGenerator.cc

```diff
@@ -175,20 +245,98 @@ G4String MuonAngleGenerator::MuonCharge() const
 
 void MuonAngleGenerator::GetDirection(G4ThreeVector& dir)
 {
-  // GetAngles from file?? Azimuth defined anticlockwise
-  // From north
-  G4double zenith  = 0.;
-  G4double azimuth = 0.;
-  distribution_->GetRandom2(azimuth, zenith);
-  // !! Current distribution in units of pi
-  zenith  *= pi;
-  azimuth *= pi;
-
-  dir.setX(sin(zenith) * sin(azimuth));
-  dir.setY(-cos(zenith));
-  dir.setZ(-sin(zenith) * cos(azimuth));
-
-  dir *= *rPhi_;
+
+  // Bool to check if zenith has a valid value. If not then resample
+  G4bool invalid_evt = true;
+
+  while(invalid_evt){
+  
+    // Amount to smear the randomly sampled zenith/azimuth values by
+    G4double zen_BW_smear{1.0e6}; 
+    G4double az_BW_smear{1.0e6};
+
+    // Generate random index weighted by the bin contents
+    G4int RN_indx = discr_dist_(RN_engine_);
+
+    // Loop over the zenith values and find the corresponding bin width to smear
+    for (G4int i = 0; i < zenith_bins_.size()-1; i++){
+        
+        // Catch very last bin
+        if (zen_BW_smear == 1e6 && i == zenith_bins_.size()-2){
+          if (zeniths_[RN_indx] >= zenith_bins_[i] 
+              && zeniths_[RN_indx] <= zenith_bins_[i+1]){
+            
+              zen_BW_smear = zen_BW_[i];
+            }
+        }
+        else {
+          if (zeniths_[RN_indx] >= zenith_bins_[i] 
+                && zeniths_[RN_indx] < zenith_bins_[i+1]){
+              
+              zen_BW_smear = zen_BW_[i];
+
+          }
+        }
+    
+    }
```

Comment:
Assuming all bins are in the input file, I would suggest to simplify the code to (same for azimuth smearing)

```suggestion
    for (G4int i = 0; i < zenith_bins_.size()-1; i++){
          if (zeniths_[RN_indx] >= zenith_bins_[i] 
              && zeniths_[RN_indx] < zenith_bins_[i+1]){
              zen_BW_smear = zen_BW_[i];
              break;
          }
```

Also, I would set the default `zen_BW_smear` and `az_BW_smear` to a meaninless negative value, say -1.

Reply:
Tidied up code by including a new function. Also set default values from 1.0e6 to std::numeric_limits<G4double>::lowest().

Kept two if statements, still need to keep the very last bin with the additional <= sign so it is included.